### DISCUSSION TITLE
Solar Powered Frog Sensor

### DISCUSSION THREAD

**Discussion by keenanjohnson** [Ribbit-Network/ribbit-network-frog-hardware#52] [Ideas]
I think it would be great from a sustainability point of view, to have a solar-powered version of the Frog Sensor!

I did some early prototyping around this concept, but I wasn't able to easily assemble a power subsystem that was reasonable to keep a Raspberry Pi operating via solar power alone.

**Primary Technical Challenge**
The main challenge is that the power draw of a Raspberry Pi (or really any Linux-based system it seems) is too high to reasonably create a small, always-on solar-powered device. I've thought of two main ways to address this:
1. Implement some system that puts the system (Wifi Radios, Processor, Sensors) to sleep and only wakes them periodically to take a measurement. This idea is great, but unfortunately, there isn't an easy way to place a Raspberry Pi in Sleep mode. One way around this would be to create additional circuitry to control the charging / powering on/off of all the electronics. This is commonly done in lots of consumer products using an additional microcontroller and in other raspberry pi carrier boards like the [Balena Fin](https://www.balena.io/fin/)
2. Switch to a microcontroller-based system like the ESP32. This would certainly make our power consumption problem much better as these microcontrollers have a proper sleep mode. However, this introduces a lot of software complexity because we will need to create our our Security / Software Update system to manage a fleet of public ESP32. Balena only works with full LInux stack systems and gives lots of nice features for free like a good VPN, and automatic software fleet updates.

**Secondary Challenges / Other Considerations** 
* Some sort of user interface/interaction will need to be designed to keep the owner of the sensor informed about the state of charge of the battery and if the sensor is receiving enough sunlight daily or if the sensor will need to be moved to a sunnier spot. This could be something as simple as an LED or could be something more expressive like a web page or information on the public dashboard.
* Assuming the solar power system uses a battery, we'll need to consider any thermal edge cases around the battery. For example, it's not recommended to charge a Lithium-Ion Cell below 0 degrees.
* Cost - The addition of a better + solar panel starts to add cost pretty quickly, so it would be great to keep the cost to a minimum if possible. Perhaps it would be cool to have the solar-powered kit be an add-on to the primary kit for those that maybe can't afford the expense initially, but would like to add it later?

**Comment by keenanjohnson** [comment]
There has been some discussion on other compute options in this issue: https://github.com/Ribbit-Network/ribbit-network-frog-sensor/issues/31

**Comment by lwbantoto** [comment]
Could we talk to some of the folks at Purple Air to see how they do things?

**Comment by keenanjohnson** [comment]
I don't believe that Purple Air offers a solar powered sensor. However, I've been told they use an ESP32.

**Comment by daveb377** [comment]
The first step is to get the power consumption down as low as possible, that makes everything easier. Sensor updates every 30 seconds, read only OS, shutting down when not on etc. Once the power consumption is low, power from a solar panel (Changing LiPo or LiFE battery)  is easier. I'm sure that there's plenty of info on the web about doing this.

**Comment by keenanjohnson** [comment]
As @djgood mentioned in #31, there is a product called the ["PiJuice"](https://github.com/PiSupply/PiJuice) that solves some of the problems. It's a small battery pack that attaches to a PI and offers a real-time clock to perform wake and sleep functions.

I actually purchased a PiJuice a long time ago and was pretty unhappy with it. However, it might still be a good starting point. Here were my notes from that time:

* The Pi Juice only seemed to be able to power a raspberry pi for around 5-6 hours at most. This really isn't long enough for a solar device to survive the night.
* The biggest issue is that if the battery dies, the Pi has to be reset by pushing a physical button, even if the battery is already recharged :(
* I had a really hard time getting the sleep function to reliably work. Sometimes I was able to get it to sleep, but not every time for some reason. This was really frustrating and I ended up returning it for other reasons before getting to the bottom of this.
* Even in sleep mode, the current was still somewhere between 50-100 mA it seemed. Not sure where the leakage was coming from and perhaps this is even a limitation of the Pi itself.
* The main issue for our application is that there was no solar panel input. [You could get a separate solar controller like this](https://www.adafruit.com/product/4755), but combined with the issue of the button reset I mentioned earlier, this doesn't really work for our purposes.

**Comment by eaudiffred** [comment]
@keenanjohnson I believe you already took some power measurements, but I had this little voltage reducer set up for something at work so I took it home and plugged it into the frog sensor for an evening.

![image](https://user-images.githubusercontent.com/88562047/138755214-2316f244-0049-4f09-aa2d-add8cca998bc.png)
![image](https://user-images.githubusercontent.com/88562047/138755247-520fb087-b1ee-47df-950c-d2c590282e85.png)

**Comment by keenanjohnson** [comment]
Thanks @eaudiffred !

**Reply by daveb377** [reply]
@eaudiffred Did you log any data? I looking at implementing a solar battery charger for this, and getting a handle on the current consumption would help a bunch.

**Comment by keenanjohnson** [comment]
In #4 we measured the power consumption to be about 2.6 watts.

**Comment by keenanjohnson** [comment]
It's my suspicion that we'll need to move to a lower power processor in order to make the power budget close for a reasonably sized solar powered sensor, but open to other opinions / more analysis :)

**Comment by eaudiffred** [comment]
@daveb377 Unfortunately I did not log any data.  I'm not sure how/if possible with that [hardware](https://www.amazon.com/DROK-Adjustable-Converter-Transformer-Protective/dp/B08RBXCJCF/ref=sr_1_8?crid=335N6AFMQRDHE&dchild=1&keywords=Drok%2BVoltage%2Breducer&qid=1635514653&sprefix=drok%2Bvoltage%2Bredu%2Caps%2C311&sr=8-8&th=1).  While it was plugged in the voltage did not move from 5.28V.  The amps fluctuated between 0.43-0.49A.  That lines up pretty closely with @keenanjohnson measurement of 2.6watts.  

If you come up with something that may work, I'd be interested in duplicating your test set up!  I was looking into batteries and solar panels from [Goal Zero](https://www.goalzero.com/) mainly because they are just plug-and-play (expensive though).  Like @keenanjohnson mentioned above, the power budget is kinda large right now and made it difficult to find something that seemed reasonable from a size and cost stand point.  Going into winter makes it harder too because the day light is disappearing fast!
